move/its/sources/service.move

```diff
@@ -273,20 +287,45 @@ module its::service {
     /// Decode an approved call and check that the source chain is trusted.
     fun decode_approved_message(self: &mut ITS, approved_message: ApprovedMessage): (String, vector<u8>) {
         let (
-            source_chain,
+            mut source_chain,
             _,
             source_address,
-            payload
+            mut payload
         ) = self.channel_mut().consume_approved_message(approved_message);
 
         assert!(self.is_trusted_address(source_chain, source_address), EUntrustedAddress);
 
+        let mut reader = abi::new_reader(payload);
+        if (reader.read_u256() == MESSAGE_TYPE_RECEIVE_FROM_HUB) {
+            assert!(source_chain.into_bytes() == ITS_HUB_CHAIN_NAME, ESenderNotHub);
+            source_chain = ascii::string(reader.read_bytes());
+            payload = reader.read_bytes();
+        };
+
         (source_chain, payload)
     }
 
     /// Send a payload to a destination chain. The destination chain needs to have a trusted address.
-    fun send_payload(self: &mut ITS, destination_chain: String, payload: vector<u8>) {
-        let destination_address = self.get_trusted_address(destination_chain);
+    fun send_payload(self: &mut ITS, mut destination_chain: String, mut payload: vector<u8>) {
+        let mut destination_address = self.get_trusted_address(destination_chain);
+
+        // Prevent sending directly to the ITS Hub chain. This is not supported yet, so fail early to prevent the user from having their funds stuck.
+        assert!(destination_chain.into_bytes() != ITS_HUB_CHAIN_NAME, EUntrustedChain);
+
+        // Check whether the ITS call should be routed via ITS hub for this destination chain
+        if(destination_address.into_bytes() == ITS_HUB_ROUTING_IDENTIFIER) {
+            let mut writter = abi::new_writer(3);
+            writter.write_u256(MESSAGE_TYPE_SEND_TO_HUB);
+            writter.write_bytes(destination_chain.into_bytes());
+            writter.write_bytes(payload);
+            payload = writter.into_bytes();
+            destination_chain = ascii::string(ITS_HUB_CHAIN_NAME);
+            destination_address = self.get_trusted_address(destination_chain);
+        };
+
+        // Check whether no trusted address was set for the destination chain
+        assert!(destination_address.length() > 0, EUntrustedChain);
```

Comment:
perhaps move this assertion to the top of the function, after the initial assignment? (if the value is the same as re-assignment value)